flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,468 @@
+
+/**
+ * Comparator for [[Row]].
+ */
+class RowComparator private (
+    /** key positions describe which fields are keys in what order */
+    val keyPositions: Array[Int],
+    /** comparators for the key fields, in the same order as the key fields */
+    val comparators: Array[TypeComparator[_]],
+    /** serializers to deserialize the first n fields for comparison */
+    val serializers: Array[TypeSerializer[_]],
+    /** auxiliary fields for normalized key support */
+    private var normalizedKeyLengths: Array[Int],
+    private var numLeadingNormalizableKeys: Int,
+    private var normalizableKeyPrefixLen: Int,
+    private var invertNormKey: Boolean)
+  extends CompositeTypeComparator[Row] with Serializable {
+
+  // null masks for serialized comparison
+  private val nullMask1 = new Array[Boolean](serializers.length)
+  private val nullMask2 = new Array[Boolean](serializers.length)
+
+  // cache for the deserialized key field objects
+  @transient
+  private var deserializedKeyFields1: Array[Any] = null
+  @transient
+  private var deserializedKeyFields2: Array[Any] = null
+
+  // null checker for reference comparison
+  private val nullChecker = new NullChecker()
+
+  def this(
+      keyPositions: Array[Int],
+      comparators: Array[TypeComparator[_]],
+      serializers: Array[TypeSerializer[_]]) = {
+    this(keyPositions, comparators, serializers, new Array[Int](keyPositions.length), 0, 0, false)
+    // set up auxiliary fields for normalized key support
+    setupAuxiliaryFields()
+  }
+
+  private def setupAuxiliaryFields(): Unit = {
+    var i = 0
+    while (i < keyPositions.length) {
+      val k: TypeComparator[_] = comparators(i)
+      // as long as the leading keys support normalized keys, we can build up the composite key
+      if (k.supportsNormalizedKey()) {
+        if (i == 0) {
+          // the first comparator decides whether we need to invert the key direction
+          invertNormKey = k.invertNormalizedKey()
+        }
+        else if (k.invertNormalizedKey() != invertNormKey) {
+          // if a successor does not agree on the inversion direction, it cannot be part of the
+          // normalized key
+          return
+        }
+        numLeadingNormalizableKeys += 1
+        val len: Int = k.getNormalizeKeyLen
+        if (len < 0) {
+          throw new RuntimeException("Comparator " + k.getClass.getName +
+            " specifies an invalid length for the normalized key: " + len)
+        }
+        normalizedKeyLengths(i) = len + 1 // add one for a null byte
+        normalizableKeyPrefixLen += len + 1 // add one for a null byte
+        if (normalizableKeyPrefixLen < 0) {
+          // overflow, which means we are out of budget for normalized key space anyways
+          normalizableKeyPrefixLen = Integer.MAX_VALUE
+          return
+        }
+      }
+      else {
+        return
+      }
+      i += 1
+    }
+  }
```

Comment:
Yes you are right. I just tried to be as close as possible to TupleComparatorBase to prevent bugs. IMHO I think my implementation is more readable as it similar to other classes (e.g. https://github.com/dataArtisans/cascading-flink/blob/master/src/main/java/com/dataartisans/flink/cascading/types/tuple/DefinedTupleComparator.java). But if other people think that your solution is better, I will change it.